tao_compiler/mlir/disc/transforms/disc_specialize_fusion_with_speculation.cc

```diff
@@ -461,48 +461,66 @@ struct DiscSpecializeFusionWithSpeculationPass
                                : 1;
       auto max_rows_per_wave =
           max_threads_per_wave / block_size * row_per_block;
-      threshold = b.create<arith::ConstantIndexOp>(loc, max_rows_per_wave);
+      Value threshold =
+          b.create<arith::ConstantIndexOp>(loc, max_rows_per_wave);
       Value operand = dominant_equivalent_op->getOperand(0);
       // #out-element is row numbers.
-      out_element_number = b.create<memref::DimOp>(loc, operand, 0);
+      Value out_element_number = b.create<memref::DimOp>(loc, operand, 0);
+
+      Value larger = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::sgt,
+                                             out_element_number, threshold);
+      Value divisible = b.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::eq,
+          b.create<arith::RemUIOp>(
+              loc, out_element_number,
+              b.create<arith::ConstantIndexOp>(loc, vector_size)),
+          b.create<arith::ConstantIndexOp>(loc, 0));
+      pred = b.create<arith::AndIOp>(loc, larger, divisible);
     } else if (fusion_type == FusionType::kLoop) {
-      threshold = b.create<arith::ConstantIndexOp>(loc, max_threads_per_wave);
       FusionPatternBase fusion_pattern(fusion_op);
       Operation* dominant_equivalent_op = fusion_pattern.getRootOps().back();
-      out_element_number =
+      Value out_element_number =
           emitNumElementsComputation(b, loc, dominant_equivalent_op);
+      if (mem_intensive_opt_experimental) {
+        vector_size = 4;
+      }
+      Value divisible = b.create<arith::CmpIOp>(
```

Comment:
Should we also check whether the buffer address divisible by vector_size?

Reply:
I see AssumeAlignmentOp inserted to annotate the buffer with the given alignment, so this check is not that critical. But since this annotation is only best-effort, it's still better to do the address alignment check if it doesn't hurt the performance.

Reply:
Very good suggestion. Checking all buffers of a fusion could be expensive, instead, I will add alignment attribute of `memref.alloc` to ensure the alignment of allocated buffers.

Reply:
That's a good idea. But I didn't see any aligned alloc implemention in RAL. I don't know if I missed something... @wyzero

Reply:
I think it’s better to use a uniform alignment  setting (e.g. a large enough setting) across all allocations if there are not so many small buffers. Mixed alignment settings may introduce redundant copies, which in turn may lead to performance drop. 

For TensorFlow users,  A default alignment value 64 is used according to following link.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/allocator.h#L114

For PyTorch users,  A default alignment value 16 or 64 is used according to following link.
https://github.com/pytorch/pytorch/blob/master/c10/core/alignment.h

For gpu users, it seems that `cudaMalloc` uses a alignment 256 by default according to following link.
https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#device-memory-accesses

Thus, I think currently we do not need to explicitly set alignment.